### PR TITLE
PAAS-2022 support CronJob

### PR DESCRIPTION
**Pull request by grosser**
tested via https://github.com/samson-test-org/example-kubernetes/pull/5

@zendesk/compute

### CODE REVIEW

File: plugins/kubernetes/app/models/kubernetes/resource.rb [hunk 1]
```diff
@@ -436,12 +436,19 @@ def request_delete
...
       def client
         @deploy_group.kubernetes_cluster.client('batch/v1')
       end
     end
 
+    class CronJob < Base
+      private
+
+      def client
+        @deploy_group.kubernetes_cluster.client('batch/v1beta1')
```
Comment (adammw) [inline plugins/kubernetes/app/models/kubernetes/resource.rb]: support 'batch/v2alpha1' for k8s <v1.8, and/or fail on older?

File: plugins/kubernetes/app/models/kubernetes/role_validator.rb [hunk 2]
```diff
@@ -187,8 +178,9 @@ def validate_containers
       primary_kinds = RoleConfigFile::PRIMARY_KINDS
       containered = templates.select { |t| primary_kinds.include?(t[:kind]) }
       containers = map_attributes([:spec, :containers], elements: containered)
-      return if containers.all? { |c| c.is_a?(Array) && c.size >= 1 }
-      @errors << "#{primary_kinds.join("/")} need at least 1 container"
+      bad = containers.select { |c| !c.is_a?(Array) || c.empty? }
+      return if bad.empty?
+      @errors << "#{containered.map { |r| r[:kind] }.join("/")} need at least 1 container"
```
Comment (adammw) [inline plugins/kubernetes/app/models/kubernetes/role_validator.rb]: nit: needs not need

File: plugins/kubernetes/app/models/kubernetes/resource.rb (hunk 1 above)
Comment (grosser) [reply plugins/kubernetes/app/models/kubernetes/resource.rb]: `You need a working Kubernetes cluster at version >= 1.8 (for CronJob). For previous versions of cluster (< 1.8) you need to explicitly enable batch/v2alpha1 API by passing --runtime-config=batch/v2alpha1=true`
... so I'd rather use the official version ... also who would run such an old cluster :trollface:

File: plugins/kubernetes/app/models/kubernetes/role_validator.rb (hunk 2 above)
Comment (grosser) [reply plugins/kubernetes/app/models/kubernetes/role_validator.rb]: could be multiple 🤷‍♂️ 
... but I guess will be 1 most of the time ...

File: plugins/kubernetes/app/models/kubernetes/resource.rb (hunk 1 above)
Comment (adammw) [reply plugins/kubernetes/app/models/kubernetes/resource.rb]: ok, but then CronJob should blow up / not validate on a <1.8 cluster.

File: plugins/kubernetes/app/models/kubernetes/role_validator.rb (hunk 2 above)
Comment (adammw) [reply plugins/kubernetes/app/models/kubernetes/role_validator.rb]: english is hard. i think it's got to do with adverb vs verb or something but :shrug:

**Comment by grosser** [comment]
deployment of it will fail just like all the other things if someone
deployed to StatefulSet a 1.1 cluster

**Comment by adammw** [comment]
fix the flaky test?

**Pull request merged**